### PR TITLE
Cleanup for 033 ver

### PR DESCRIPTION
**Pull request by sander-visser**
Fixed 1 old error (in0.3.2), one new error (in my last optimization) + added another level of optimizations...

Prepared for 0.3.3 tagging

### CODE REVIEW

File: tests/expected_gen/ApxNode_Test.c [hunk 1]
```diff
@@ -172,17 +187,40 @@ Std_ReturnType ApxNode_Write_Test_PU32Port(uint32 val)
...
-   (void)offset;
-   (void)len;
+
+   if (offset < APX_RX_OFFSET_RU8PORT)
+   {
+      offset = APX_RX_OFFSET_RU8PORT;
+   }
+   while (offset < endOffset)
+   {
+      switch(offset)
```
Comment (sander-visser) [inline tests/expected_gen/ApxNode_Test.c]: Only create switch statements between the first and last callback function

File: apx/__init__.py [hunk 2]
```diff
@@ -1,4 +1,4 @@
-__version__ = "0.3.2"
+__version__ = "0.3.3"
```
Comment (sander-visser) [inline apx/__init__.py]: Maybe the version used to generate a file should be shown in the c file as a comment along with a remark that the file is auto generated?

Comment (sander-visser) [reply apx/__init__.py]: Implemented that feature...

File: tests/expected_gen/ApxNode_Test.c [hunk 3]
```diff
@@ -172,17 +187,40 @@ Std_ReturnType ApxNode_Write_Test_PU32Port(uint32 val)
 
 void Test_inPortDataWritten(void *arg, apx_nodeData_t *nodeData, uint32_t offset, uint32_t len)
 {
+   union data_tag
```
Comment (sander-visser) [inline tests/expected_gen/ApxNode_Test.c]: C89 compilation issue if (void)arg is before the declarations @ovenystas

File: tests/expected_gen/ApxNode_Test.c (hunk 1 above)
Comment (cogu) [reply tests/expected_gen/ApxNode_Test.c]: Current code generator is not super efficient. However, it handles all situations.
For example, if entire file has been written to (which happens the first time during connect) it triggers all callbacks.
It also handles the case where just one port has been written to. A more efficient code generator needs to also solve both problems.

File: tests/generator_test.py [hunk 4]
```diff
@@ -34,12 +34,14 @@ def test_code_generator(self):
         node.append(apx.ProvidePort('PU16ARPort','S[4]','={65535, 65535, 65535, 65535}'))
         node.append(apx.ProvidePort('PU32Port','L','=4294967295'))
 
+        callback_map = { 'RU8Port': 'RU8Port_cb_func' }
```
Comment (cogu) [inline tests/generator_test.py]: No, you must create an APX node with at least 3 callbacks and prove that all callbacks are called when if entire file is written.

File: tests/expected_gen/ApxNode_Test.c (hunk 1 above)
Comment (sander-visser) [reply tests/expected_gen/ApxNode_Test.c]: The new generator can also handle triggering all callbacks when full file is written. It just starts the trigger port search at the first callback and ends at the last. There is still potential to optimize sparse callbacks inbetween first and last if the node is not created with all callbacks sorted first or last...

File: tests/generator_test.py (hunk 4 above)
Comment (sander-visser) [reply tests/generator_test.py]: Added unittest showing it works when entire file is writte

File: tests/expected_gen/ApxNode_Test.c [hunk 5]
```diff
@@ -42,8 +59,8 @@ static const char *m_apxDefinitionData=
 "P\"PU32Port\"L:=4294967295\n"
 "R\"RS16ARPort\"s[3]:={32767, 1, 0}\n"
 "R\"RS32Port\"l:=1\n"
```
Comment (sander-visser) [inline tests/expected_gen/ApxNode_Test.c]: finalized the node before calling the generator to test generating unordered ports

File: tests/expected_gen/ApxNode_TestCallbackLast.c [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+   union data_tag
+   {
+      uint8 RU8Port;
+   } data;
+   uint32_t endOffset = offset + len;
+
+   (void)arg;
+   (void)nodeData;
+
+   if (offset < APX_RX_OFFSET_RU8PORT)
```
Comment (sander-visser) [inline tests/expected_gen/ApxNode_TestCallbackLast.c]: skipping ahead to first port with callback

File: tests/expected_gen/ApxNode_TestCallbackFirst.c [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+   while (offset < endOffset)
+   {
+      switch(offset)
+      {
+      case APX_RX_OFFSET_RS32PORT:
+         (void) ApxNode_Read_TestCallbackFirst_RS32Port(&data.RS32Port);
+         RS32Port_cb_func(data.RS32Port);
+         offset += APX_RX_LEN_RS32PORT;
+         break;
+      //case APX_RX_OFFSET_RU8PORT:
```
Comment (sander-visser) [inline tests/expected_gen/ApxNode_TestCallbackFirst.c]: after the last callback has been processed skip all the way to end

**Comment by sander-visser** [comment]
Done updating pullreq according to comments

**Comment by sander-visser** [comment]
> 
> 
> Change unit test to include 3 callbacks in the generated C code.

done

**Comment by sander-visser** [comment]
kj

File: tests/expected_gen/ApxNode_TestCallbackLast.c (hunk 6 above)
Comment (cogu) [reply tests/expected_gen/ApxNode_TestCallbackLast.c]: Great! Now even I understand what is going on. :)

**Comment by cogu** [comment]
Great work!

**Pull request merged**